pkg/strategy/grid/strategy.go

```diff
@@ -0,0 +1,274 @@
+			Symbol:      s.Symbol,
+			Side:        types.SideTypeSell,
+			Type:        types.OrderTypeLimit,
+			Market:      s.Market,
+			Quantity:    s.BaseQuantity,
+			Price:       startPrice,
+			TimeInForce: "GTC",
+		})
+
+		startPrice += s.GridPips.Float64()
+	}
+
+	orders, err := orderExecutor.SubmitOrders(context.Background(), submitOrders...)
+	if err != nil {
+		log.WithError(err).Error("submit ask order error")
+		return
+	}
+
+	s.mu.Lock()
+	for i := range orders {
+		var order = orders[i]
+		log.Infof("adding order %d to the active ask order pool...", order.OrderID)
+		s.activeAskOrders[order.OrderID] = order
+	}
+	s.mu.Unlock()
+}
+
+func (s *Strategy) updateOrders(orderExecutor bbgo.OrderExecutor, session *bbgo.ExchangeSession) {
+	log.Infof("checking grid orders, bids=%d asks=%d", len(s.activeBidOrders), len(s.activeAskOrders))
+
+	for _, o := range s.activeBidOrders {
+		log.Infof("bid order: %d -> %s", o.OrderID, o.Status)
+	}
+
+	for _, o := range s.activeAskOrders {
+		log.Infof("ask order: %d -> %s", o.OrderID, o.Status)
+	}
+
+	if len(s.activeBidOrders) < s.GridNum {
+		log.Infof("active bid orders not enough: %d < %d, updating...", len(s.activeBidOrders), s.GridNum)
+		s.updateBidOrders(orderExecutor, session)
+	}
+
+	if len(s.activeAskOrders) < s.GridNum {
+		log.Infof("active ask orders not enough: %d < %d, updating...", len(s.activeAskOrders), s.GridNum)
+		s.updateAskOrders(orderExecutor, session)
+	}
+}
+
+func (s *Strategy) Run(ctx context.Context, orderExecutor bbgo.OrderExecutor, session *bbgo.ExchangeSession) error {
+	if s.GridNum == 0 {
+		s.GridNum = 2
+	}
+
+	s.boll = s.StandardIndicatorSet.GetBOLL(types.IntervalWindow{
+		Interval: s.Interval,
+		Window:   21,
+	})
+
+	// we don't persist orders so that we can not clear the previous orders for now. just need time to support this.
+	// TODO: pull this map out and add mutex lock
+	s.activeBidOrders = make(map[uint64]types.Order)
+	s.activeAskOrders = make(map[uint64]types.Order)
+
+	session.Stream.OnOrderUpdate(func(order types.Order) {
+		log.Infof("received order update: %+v", order)
+
+		if order.Symbol != s.Symbol {
+			return
+		}
+
+		s.mu.Lock()
+		defer s.mu.Unlock()
+
+		switch order.Status {
+
+		case types.OrderStatusCanceled, types.OrderStatusRejected, types.OrderStatusFilled:
+			log.Infof("order status %s, removing %d from the active order pool...", order.Status, order.OrderID)
+
+			switch order.Side {
+			case types.SideTypeSell:
+				delete(s.activeAskOrders, order.OrderID)
+			case types.SideTypeBuy:
+				delete(s.activeBidOrders, order.OrderID)
+
+			}
+
+		default:
+			log.Infof("order status %s, updating %d to the active order pool...", order.Status, order.OrderID)
+			switch order.Side {
+			case types.SideTypeSell:
+				s.activeAskOrders[order.OrderID] = order
+			case types.SideTypeBuy:
+				s.activeBidOrders[order.OrderID] = order
+			}
+		}
+	})
+
+	go func() {
+		ticker := time.NewTicker(10 * time.Second)
```

Comment:
Could we increase this ticker period to be greater than normal network timeout, like 1 minute?
It's just a rare corner case. If an order update comes late due to network delay, it is likely to place duplicated orders.
However, setting to 1 minute does NOT completely fix this but it becomes much rarer :)

Reply:
surely!

Reply:
fixed